controllers/service_cache.go

```diff
@@ -0,0 +1,122 @@
+package controllers
+
+import (
+	"context"
+	stderrors "errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"slime.io/slime/framework/util"
+)
+
+func newSvcCache(clientSet *kubernetes.Clientset) (*NsSvcCache, *LabelSvcCache, error) {
+	log := log.WithField("function", "newLabelSvcCache")
+	nsSvcCache := &NsSvcCache{Data: map[string]map[string]struct{}{}}
+	labelSvcCache := &LabelSvcCache{Data: map[LabelItem]map[string]struct{}{}}
+
+	// init labelSvcCache
+	services, err := clientSet.CoreV1().Services("").List(metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, stderrors.New("failed to get service list")
+	}
+
+	for _, service := range services.Items {
+		ns := service.GetNamespace()
+		name := service.GetName()
+		svc := ns + "/" + name
```

Comment:
we can also use `NamespaceName` for svc description to reduce str-concat cost as an optimize case later.